pkg/sql/logictest/tests/cockroach-go-testserver-24.3/cockroach-go-testserver-24_3_test: TestLogic_upgrade failed

pkg/sql/logictest/tests/cockroach-go-testserver-24.3/cockroach-go-testserver-24_3_test.TestLogic_upgrade [failed](https://tanzanite.cluster.engflow.com/invocations/default/6d906e79-9f1c-4338-92cc-c8266d86a8c3?testReportRun=5&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvc3FsL2xvZ2ljdGVzdC90ZXN0cy9jb2Nrcm9hY2gtZ28tdGVzdHNlcnZlci0yNC4zOmNvY2tyb2FjaC1nby10ZXN0c2VydmVyLTI0XzNfdGVzdA==) on release-25.1 @ [29c0a3d5161b6818e414efc9c35321c1c0416e19](https://github.com/cockroachdb/cockroach/commits/29c0a3d5161b6818e414efc9c35321c1c0416e19):

Fatal error:

```
panic: test timed out after 15m0s
running tests:
	TestLogic_upgrade (15m0s)

```
Stack: 

```
goroutine 92 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x385
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x2d
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestLogic_upgrade
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestLogic_upgrade1286049234
    test_log_scope.go:76: use -show-logs to present logs inline
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=5</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #139100 pkg/sql/logictest/tests/cockroach-go-testserver-24.3/cockroach-go-testserver-24_3_test: TestLogic_upgrade failed [C-test-failure O-robot T-sql-foundations branch-master]
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLogic_upgrade.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-46634

pkg/sql/logictest/tests/cockroach-go-testserver-24.3/cockroach-go-testserver-24_3_test.TestLogic_upgrade [failed](https://tanzanite.cluster.engflow.com/invocations/default/4fd767b6-20ce-460f-971a-d91ef8e4408a?testReportRun=10&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvc3FsL2xvZ2ljdGVzdC90ZXN0cy9jb2Nrcm9hY2gtZ28tdGVzdHNlcnZlci0yNC4zOmNvY2tyb2FjaC1nby10ZXN0c2VydmVyLTI0XzNfdGVzdA==) on release-25.1 @ [4bb718e70117615cd5ac3210ae9dcfcd67097ab7](https://github.com/cockroachdb/cockroach/commits/4bb718e70117615cd5ac3210ae9dcfcd67097ab7):

Fatal error:

```
panic: test timed out after 15m0s
running tests:
	TestLogic_upgrade (15m0s)

```
Stack: 

```
goroutine 115 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x385
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x2d
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestLogic_upgrade
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestLogic_upgrade2180664852
    test_log_scope.go:76: use -show-logs to present logs inline
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=10</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #139100 pkg/sql/logictest/tests/cockroach-go-testserver-24.3/cockroach-go-testserver-24_3_test: TestLogic_upgrade failed [C-test-failure O-robot P-1 T-sql-foundations branch-master]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLogic_upgrade.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/server/server_sql.go
+++ b/pkg/server/server_sql.go
@@ -1643,7 +1643,7 @@ func (s *SQLServer) preStart(
 	// care about what uses the SQL server before those migrations
 	// run.
 
-	s.leaseMgr.RefreshLeases(ctx, stopper, s.execCfg.DB)
+	s.leaseMgr.StartRefreshLeasesTask(ctx, stopper, s.execCfg.DB)
 	s.leaseMgr.RunBackgroundLeasingTask(ctx)
 
 	if err := s.jobRegistry.Start(ctx, stopper); err != nil {

--- a/pkg/sql/catalog/lease/lease.go
+++ b/pkg/sql/catalog/lease/lease.go
@@ -1154,6 +1154,10 @@ type Manager struct {
 	// detected by the lease manager. Once this count is incremented new data
 	// is available.
 	leaseGeneration atomic.Int64
+
+	// waitForInit used when the lease manager is starting up prevent leases from
+	// being acquired before the range feed.
+	waitForInit chan struct{}
 }
 
 const leaseConcurrencyLimit = 5
@@ -1248,7 +1252,8 @@ func NewLeaseManager(
 	lm.storage.writer = newKVWriter(codec, db.KV(), keys.LeaseTableID, settingsWatcher, lm)
 	lm.stopper.AddCloser(lm.sem.Closer("stopper"))
 	lm.mu.descriptors = make(map[descpb.ID]*descriptorState)
-	// We are going to start the range feed later when RefreshLeases
+	lm.waitForInit = make(chan struct{})
+	// We are going to start the range feed later when StartRefreshLeasesTask
 	// is invoked inside pre-start. So, that guarantees all range feed events
 	// that will be generated will be after the current time. So, historical
 	// queries with in this tenant (i.e. PCR catalog reader) before this point are
@@ -1603,8 +1608,17 @@ func (m *Manager) isDescriptorStateEmpty(id descpb.ID) bool {
 	return len(st.mu.active.data) == 0
 }
 
+// maybeWaitForInit waits for the lease manager to startup.
+func (m *Manager) maybeWaitForInit() {
+	select {
+	case <-m.waitForInit:
+	case <-m.stopper.ShouldQuiesce():
+	}
+}
+
 // If create is set, cache and stopper need to be set as well.
 func (m *Manager) findDescriptorState(id descpb.ID, create bool) *descriptorState {
+	m.maybeWaitForInit()
 	m.mu.Lock()
 	defer m.mu.Unlock()
 	t := m.mu.descriptors[id]
@@ -1615,13 +1629,14 @@ func (m *Manager) findDescriptorState(id descpb.ID, create bool) *descriptorStat
 	return t
 }
 
-// RefreshLeases starts a goroutine that refreshes the lease manager
+// StartRefreshLeasesTask starts a goroutine that refreshes the lease manager
 // leases for descriptors received in the latest system configuration via gossip or
 // rangefeeds. This function must be passed a non-nil gossip if
 // RangefeedLeases is not active.
-func (m *Manager) RefreshLeases(ctx context.Context, s *stop.Stopper, db *kv.DB) {
+func (m *Manager) StartRefreshLeasesTask(ctx context.Context, s *stop.Stopper, db *kv.DB) {
 	m.mu.Lock()
 	defer m.mu.Unlock()
+	defer close(m.waitForInit)
 	m.watchForUpdates(ctx)
 	_ = s.RunAsyncTask(ctx, "refresh-leases", func(ctx context.Context) {
 		for {
@@ -2409,4 +2424,12 @@ func (m *Manager) TestingSetDisableRangeFeedCheckpointFn(disable bool) chan stru
 		m.testingKnobs.RangeFeedResetChannel = nil
 	}
 	return m.testingKnobs.RangeFeedResetChannel
+}
+
+// TestingMarkInit marks the lease manager as initialized without a range feed being started.
+// This is only used for testing.
+func (m *Manager) TestingMarkInit() {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	close(m.waitForInit)
 }

--- a/pkg/sql/catalog/lease/lease_test.go
+++ b/pkg/sql/catalog/lease/lease_test.go
@@ -283,6 +283,7 @@ func (t *leaseTest) node(nodeID uint32) *lease.Manager {
 		)
 		ctx := logtags.AddTag(context.Background(), "leasemgr", nodeID)
 		mgr.RunBackgroundLeasingTask(ctx)
+		mgr.TestingMarkInit()
 		t.nodes[nodeID] = mgr
 	}
 	return mgr

